assets/src/edit-story/app/history/historyProvider.js

```diff
@@ -36,11 +37,14 @@ function HistoryProvider({ children, size }) {
     undo,
     redo,
   } = useHistoryReducer(size);
+  const canUndo = offset < historyLength - 1;
+
+  usePreventWindowUnload(canUndo);
```

Comment:
The `canUndo` survives editor savings. Thus, I don't think it can be used here - it will be regularly overwriting the `usePreventWindowUnload(false)` set in the save story.

Reply:
It only will override the `usePreventWindowUnload(false)` set in the save story action when you have something new in history. But, as @spacedmonkey reported, there are specific cases that we have "hidden" history changes that can in fact create side effects. 

That being said, I am a bit in doubt about how to track truly canvas changes since the history is not trustable. Maybe have a dedicated action for it? E.g. `SET_PREVENT_WINDOW_UNLOAD` and `REMOVE_PREVENT_WINDOW_UNLOAD`. So we can call it directly when we want to enable or disable instead of applying a implicit controller. Thoughts? Do you guys have another suggestion?

Reply:
I think what you're saying, if the case, is a separate issue for history where it should not have "hidden" updates. However, my point above still stands `canUndo` will almost always be `true`: including for a just saved story. That means, at any random time, this code might execute `usePreventWindowUnload(true)`, including just after saving a story. What we want to do here, `usePreventWindowUnload(hasChangedSinceLastSave)`. If we had it here, you wouldn't need to call `usePreventWindowUnload` from anywhere else.

Reply:
Sure! I'm working on it. Just not sure yet from where I will extract a trustable history changes. My best guess is still `historyLength` :)